draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,978 @@
+
+1.  MUST verify the `desc:handoff-receive` (see `desc:handoff-receive` section).
+    If it is incorrect, abort the handoff; otherwise, continue.
+3.  If the gift has already been deposited return the gift; otherwise return the
+    gift when it is received.
+
+Note: Gifts are per-session. This means that gifts deposited within the `Gifter
+<-> Exporter` session should only be available for withdrawal within the
+`Receiver <-> Exporter session`. The receiver should be the only party able to
+withdraw the gift left by the gifter. Implementers MUST ensure that the
+management of gifts adheres to this per-session requirement, preventing
+unauthorized access to gifts.
+
+# Operations
+
+## [`op:start-session`](#op-start-session)
+
+On a new connection, a key pair for this session should be generated. This
+key pair should be an EdDSA key pair with a SHA512 hash.
+
+This operation is used when a new connection is initiated over CapTP. Both
+parties MUST send upon a new connection. The operation looks like this:
+
+```text
+<op:start-session [captp-version              ; String value
+                   session-pubkey             ; CapTP public key value
+                   acceptable-location        ; OCapN Reference type
+                   acceptable-location-sig]>  ; CapTP signature
+```
+
+It's important that we only have one bidirectional connection between a CapTP
+session. Before trying to connect to a machine, it's important to check that a
+connection is not already open before proceeding. There is a race condition
+called the "crossed hellos" problem whereby each side tries to open a new
+connection at the same time and only one of these connection attempts must
+succeed. The way to check which attempt should succeed is to sort the serialized
+`session-pubkey`, the highest of the two wins.
+
+### Constructing and sending
+
+The `captp-version` MUST be `1.0`.
+
+The `session-pubkey` is the public key part of the per-session key pair
+generated for this connection. This is serialized in accordance with
+[Cryptography](#cryptography)
+
+The `acceptable-location` is a OCapN Locator which represents the location where
+the session is accessable.
+
+The `acceptable-location-sig` is the signature of the serialized
+`acceptable-location`. The signature is created using the private key from the
+per-session key pair. This is serialized in accordance with
+[Cryptography](#cryptography)
+
+### Receiving
+
+The `captp-version` MUST be equal to `1.0`. If the version does not match, the
+connection MUST be aborted.
+
+The `acceptable-location-sig` MUST be valid that the `session-pubkey` provided a
+valid signature of `acceptable-location`.
+
+## [`op:bootstrap`](#op-bootstrap)
+
+This operation requests the bootstrap object to be available at the specified
+`answer-position`. The message looks like:
+
+```text
+<op:bootstrap [answer-position    ; positive integer or 0
+               resolve-me-desc]>  ; desc:import-object | desc:import-promise
```

Comment:
resolve-me-desc here and in op:deliver seem to be redundant functionality of what op:listen does. See #15 for why.